### PR TITLE
Stepper: Update calypso_signup_start tracking

### PR DESCRIPTION
**Pull request by edanzer**
### Proposed Changes

This PR makes a number of changes to when and how we track calypso_signup_start for stepper-based flows. It mostly tries to implement feedback from martech/analytics on this P2: pau2Xa-4DX-p2

**Changes include:**
   - Do not call `recordTracksEvent( 'calypso_signup_start')` directly. Rather call `recordSignupStart()` from `client/lib/analytics`. This method triggers other important tracking. 
   - Rather than trigger tracking from every flow separately, trigger it once from the shared `stepper/internals/index.tsx` file. This reduces duplication and allows us to tweak logic and conditions without duplicating across files.
   - We currently trigger calypso_signup_start every time someone re-enters stepper (going to wp admin or start page and coming back trigger another calypso_signup_start). This fixes that by checking stepProgress and only triggering signup_start tracking on the first step of each stepper flow (stepProgress.progress === 0).

### Testing Instructions

**Testing Time: Medium**
**Review Time: Medium**

**Preparation**. You will need to be able to see tracks events in order to test. You have two options: 
- Option 1 (recommended): Install and use the [Tracks Vigilante](https://github.com/Automattic/tracks-chrome-extension) Chrome extension.
- Option 2: Open Developer Tools > Network. Filter requests with `t.gif`. You'll find tracks event meta under payload for each request. 

**Testing**.
1. Checkout this branch and run yarn and yarn start if needed. 
2. Clear tracks events from tracks vigilante or from your dev console. 
3. Test Free Flow: Go to http://calypso.localhost:3000/setup/free/intro and confirm calypso_signup_start fires. Go through other steps of Free flow to launchpad, and confirm calypso_signup_start does not fires again.
4. Test Link in Bio: Clear tracks events. Go to http://calypso.localhost:3000/setup/link-in-bio/intro and confirm calypso_signup_start fires. Go through other steps of Free flow to launchpad, and confirm calypso_signup_start does not fires again.
5. Test Newsletter: Clear tracks events. Go to http://calypso.localhost:3000/setup/newsletter/intro and confirm calypso_signup_start fires. Go through other steps of Free flow to launchpad, and confirm calypso_signup_start does not fires again.
6. Test VideoPress: Clear tracks events. Go to http://calypso.localhost:3000/setup/videopress/intro and confirm calypso_signup_start fires. Go through other steps of Free flow to launchpad, and confirm calypso_signup_start does not fires again.
7. Test Sensei: Clear tracks events. Go to http://calypso.localhost:3000/setup/sensei and confirm calypso_signup_start fires. Go through other steps of Free flow to launchpad, and confirm calypso_signup_start does not fires again.

Related to https://github.com/Automattic/wp-calypso/issues/70883

### CODE REVIEW

**Comment by agrullon95** [comment]
@edanzer - If the entry point for the free flow leads to `/setup/free/freeSetup` we won't be able to use `stepProgress?.progress === 0` to properly trigger the `calypso_signup_start` tracking event.

File: client/landing/stepper/declarative-flow/internals/index.tsx [hunk 1]
```diff
@@ -75,6 +77,14 @@ export const FlowRenderer: React.FC< { flow: Flow } > = ( { flow } ) => {
 		window.scrollTo( 0, 0 );
 	}, [ location ] );
 
+	useEffect( () => {
+		if ( flow && stepProgress !== undefined && stepProgress?.progress === 0 ) {
+			const ref = new URLSearchParams( window.location.search ).get( 'ref' ) || '';
```
Comment (bangank36) [inline client/landing/stepper/declarative-flow/internals/index.tsx]: I think you can use `useQuery` hook, similar to https://github.com/Automattic/wp-calypso/blob/1b2690b821df5fde4b6b497054fd77fbb6dcaf6d/client/landing/stepper/declarative-flow/copy-site.tsx#L31

**Comment by jeyip** [comment]
## Testing
### Requirements
- [x] Free Flow: confirm calypso_signup_start fires once
- [x] Link in Bio Flow: confirm calypso_signup_start fires once
- [ ] Newsletter Flow: confirm calypso_signup_start fires once
- [x] VideoPress Flow: confirm calypso_signup_start fires once
- [x] Sensei Flow: confirm calypso_signup_start fires once

### Browsers
- [x] Chrome

I'm seeing `calypso_signup_start` triggered twice for Newsletter flows

File: client/landing/stepper/declarative-flow/internals/index.tsx [hunk 2]
```diff
@@ -2,14 +2,17 @@ import { ProgressBar } from '@automattic/components';
...
 import Modal from 'react-modal';
 import { Switch, Route, Redirect, generatePath, useHistory, useLocation } from 'react-router-dom';
 import DocumentHead from 'calypso/components/data/document-head';
 import WordPressLogo from 'calypso/components/wordpress-logo';
+import { useQuery } from 'calypso/landing/stepper/hooks/use-query';
 import { STEPPER_INTERNAL_STORE } from 'calypso/landing/stepper/stores';
+import { recordFullStoryEvent } from 'calypso/lib/analytics/fullstory';
 import { recordPageView } from 'calypso/lib/analytics/page-view';
 import SignupHeader from 'calypso/signup/signup-header';
+import { recordSignupStart } from '../../../../lib/analytics/signup';
```
Comment (jeyip) [inline client/landing/stepper/declarative-flow/internals/index.tsx]: `NB`: 

Could we use `calypso/lib/analytics/signup` ?

**Comment by jeyip** [comment]
> I'm seeing calypso_signup_start triggered twice for Newsletter flows

Fine to address in a follow-up since what you’re changing noticeably improves the current situation

**Review by jeyip — APPROVED** [review]
Thanks for taking the time to fix our data tracking -- it's particularly important to get this right to understand our bottom-line impact with the Launchpad

**Comment by jeyip** [comment]
If possible, I'd also very much encourage a unit test for the new tracking logic that we're adding to `client/landing/stepper/declarative-flow/internals/index.tsx`. If that makes sense and you want help with this, let me know 👍

File: client/landing/stepper/declarative-flow/internals/index.tsx (hunk 1 above)
Comment (edanzer) [reply client/landing/stepper/declarative-flow/internals/index.tsx]: Thanks @bangank36! Updated this.

File: client/landing/stepper/declarative-flow/internals/index.tsx (hunk 2 above)
Comment (edanzer) [reply client/landing/stepper/declarative-flow/internals/index.tsx]: Good catch and much simpler. Updated!

**Comment by edanzer** [comment]
> > I'm seeing calypso_signup_start triggered twice for Newsletter flows
> 
> Fine to address in a follow-up since what you’re changing noticeably improves the current situation

@jeyip Good catch. Just looking at this, the issue is that the newsletter still uses the old /start/ signup pages. Those are firing an extra start event when the domain. The relevant line of code is https://github.com/Automattic/wp-calypso/blob/41b70766845caf31f412248746085325006fe0d3/client/signup/main.jsx#L258 - if I comment that out, it goes away. So I think we just need to wrap that in a conditional. As this PR is already getting large, and that touches an extra area of code, I'll probably leave that for a follow up.

**Comment by agrullon95** [comment]
~~Im noticing that the `calypso_signup_start` event is triggered twice when I visit the videopress intro page.
Also, i'm not able to reach the sensei flow. I'm redirected to the calypso home.~~

I'm no longer seeing the video press event issue and Im able to load the sensei flow correctly now. I do see that the newsletter flow is firing the event twice, but it seems like you're going to address that in a follow up PR.

### Tests:

- [x] Test Free Flow: Go to http://calypso.localhost:3000/setup/free/intro and confirm calypso_signup_start fires. Go through other steps of Free flow to launchpad, and confirm calypso_signup_start does not fires again.
- [x] Test Link in Bio: Clear tracks events. Go to http://calypso.localhost:3000/setup/link-in-bio/intro and confirm calypso_signup_start fires. Go through other steps of Free flow to launchpad, and confirm calypso_signup_start does not fires again.
- [ ] Test Newsletter: Clear tracks events. Go to http://calypso.localhost:3000/setup/newsletter/intro and confirm calypso_signup_start fires. Go through other steps of Free flow to launchpad, and confirm calypso_signup_start does not fires again.
- [x] Test VideoPress: Clear tracks events. Go to http://calypso.localhost:3000/setup/videopress/intro and confirm calypso_signup_start fires. Go through other steps of Free flow to launchpad, and confirm calypso_signup_start does not fires again.
- [x] Test Sensei: Clear tracks events. Go to http://calypso.localhost:3000/setup/sensei and confirm calypso_signup_start fires. Go through other steps of Free flow to launchpad, and confirm calypso_signup_start does not fires again.

**Review by agrullon95 — APPROVED** [review]

**Pull request merged**